# Shirley's work

In [ ]:
!pip install sentence-transformers

In [ ]:
from sentence_transformers import SentenceTransformer, util

model = SentenceTransformer('all-MiniLM-L6-v2')

persona_data = {
    "Name": "Yuyang Zeng",
    "Age": 24,
    "Occupation": "UofT Engineering Student",
    "Gender": "Male",
    "Interest": ["Basketball", "Soccer", "Fashion", "Travel", "Technology"],
    "Recent Searches": [
        "TV stand",
        "Clorox Disinfecting Wipes",
        "Desk Clamp Power Bar",
        "Durex Condoms",
        "Temporary Hair Colour Wax",
        "Stanley IceFlow Stainless Steel"
    ]
}

persona_text = (
    f"Name: {persona_data['Name']}, Age: {persona_data['Age']}, Occupation: {persona_data['Occupation']}, "
    f"Gender: {persona_data['Gender']}. Interests include {', '.join(persona_data['Interest'])}. "
    f"Recent searches include {', '.join(persona_data['Recent Searches'])}."
)

generated_queries = [
    "What are the best AI tools for developers?",
    "Latest smartphones with AI-powered cameras",
    "Top recipes for cooking with spices",
    "New fitness gadgets for health enthusiasts",
    "AI advancements in robotics",
    "Best travel destinations for students",
    "Where to buy basketball shoes online",
    "Fashion trends for the summer season",
    "Best disinfecting wipes for home use",
    "How to use hair colour wax safely"
]

persona_embedding = model.encode(persona_text, convert_to_tensor=True)

print("\nQuery Relevance Scores:")
for query in generated_queries:
    query_embedding = model.encode(query, convert_to_tensor=True)

    similarity_score = util.cos_sim(persona_embedding, query_embedding).item()

    print(f"Query: '{query}'\nRelevance Score: {similarity_score:.2f}\n")


/usr/local/lib/python3.10/dist-packages/sentence_transformers/cross_encoder/CrossEncoder.py:13: TqdmExperimentalWarning: Using `tqdm.autonotebook.tqdm` in notebook mode. Use `tqdm.tqdm` instead to force console mode (e.g. in jupyter console)
  from tqdm.autonotebook import tqdm, trange
/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


modules.json:   0%|          | 0.00/349 [00:00<?, ?B/s]

config_sentence_transformers.json:   0%|          | 0.00/116 [00:00<?, ?B/s]

README.md:   0%|          | 0.00/10.7k [00:00<?, ?B/s]

sentence_bert_config.json:   0%|          | 0.00/53.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/612 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/90.9M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/350 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/466k [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/112 [00:00<?, ?B/s]

1_Pooling/config.json:   0%|          | 0.00/190 [00:00<?, ?B/s]


Query Relevance Scores:
Query: 'What are the best AI tools for developers?'
Relevance Score: 0.08

Query: 'Latest smartphones with AI-powered cameras'
Relevance Score: 0.03

Query: 'Top recipes for cooking with spices'
Relevance Score: 0.08

Query: 'New fitness gadgets for health enthusiasts'
Relevance Score: 0.24

Query: 'AI advancements in robotics'
Relevance Score: 0.09

Query: 'Best travel destinations for students'
Relevance Score: 0.14

Query: 'Where to buy basketball shoes online'
Relevance Score: 0.09

Query: 'Fashion trends for the summer season'
Relevance Score: 0.13

Query: 'Best disinfecting wipes for home use'
Relevance Score: 0.21

Query: 'How to use hair colour wax safely'
Relevance Score: 0.14



In [ ]:
from sentence_transformers import SentenceTransformer, util

# model = SentenceTransformer('all-MiniLM-L6-v2')
# model = SentenceTransformer('all-distilroberta-v1')
model = SentenceTransformer('paraphrase-MiniLM-L12-v2')


persona_data = {
    "Name": "Yuyang Zeng",
    "Age": 24,
    "Occupation": "UofT Engineering Student",
    "Gender": "Male",
    "Interest": ["Basketball", "Soccer", "Fashion", "Travel", "Technology"],
    "Recent Searches": [
        "TV stand",
        "Clorox Disinfecting Wipes",
        "Desk Clamp Power Bar",
        "Durex Condoms",
        "Temporary Hair Colour Wax",
        "Stanley IceFlow Stainless Steel"
    ]
}

# persona_text = (
#     f"Name: {persona_data['Name']}, Age: {persona_data['Age']}, Occupation: {persona_data['Occupation']}, "
#     f"Gender: {persona_data['Gender']}. Interests include {', '.join(persona_data['Interest'])}. "
#     f"Recent searches include {', '.join(persona_data['Recent Searches'])}."
# )
# persona_text = (
#     "Yuyang Zeng is a 24-year-old computer engineering student at UofT who is interested in basketball, soccer, "
#     "and staying up to date with the latest fashion trends. Yuyang enjoys traveling to new places and "
#     "exploring technology, especially the newest gadgets and innovations. Recently, Yuyang has been searching "
#     "for items like TV stands, disinfecting wipes, power bars for his desk setup, and hair color wax for a change of style."
# )
persona_text = (
    "Yuyang Zeng is a 24-year-old computer engineering student at UofT. "
    "He is interested in artificial intelligence, software development, and exploring new technologies. "
    "Yuyang frequently researches AI tools and techniques to enhance his development skills."
)



# generated_queries = [
#     "What are the best AI tools for developers?",
#     "Latest smartphones with AI-powered cameras",
#     "Top recipes for cooking with spices",
#     "New fitness gadgets for health enthusiasts",
#     "AI advancements in robotics",
#     "Best travel destinations for students",
#     "Where to buy basketball shoes online",
#     "Fashion trends for the summer season",
#     "Best disinfecting wipes for home use",
#     "How to use hair colour wax safely"
# ]
generated_queries = [
    "What are the best AI tools for developers?",
]

persona_embedding = model.encode(persona_text, convert_to_tensor=True)

print("\nQuery Relevance Scores:")
for query in generated_queries:
    query_embedding = model.encode(query, convert_to_tensor=True)

    similarity_score = util.cos_sim(persona_embedding, query_embedding).item()

    print(f"Query: '{query}'\nRelevance Score: {similarity_score:.2f}\n")


modules.json:   0%|          | 0.00/229 [00:00<?, ?B/s]

config_sentence_transformers.json:   0%|          | 0.00/122 [00:00<?, ?B/s]

README.md:   0%|          | 0.00/3.73k [00:00<?, ?B/s]

sentence_bert_config.json:   0%|          | 0.00/53.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/631 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/133M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/316 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/466k [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/112 [00:00<?, ?B/s]

1_Pooling/config.json:   0%|          | 0.00/190 [00:00<?, ?B/s]


Query Relevance Scores:
Query: 'What are the best AI tools for developers?'
Relevance Score: 0.28



sentence bert relevance scores are not satisfactory, as the persona and query are relevant but the score is very low.
Therefore, not a valid way of evaluating our output

Exploring with BERTScore

In [ ]:
!pip install bert-score


   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 61.1/61.1 kB 3.5 MB/s eta 0:00:00


In [ ]:

from bert_score import score

persona_text = (
    "Yuyang Zeng is a 24-year-old computer engineering student at UofT. "
    "He is interested in artificial intelligence, software development, and exploring new technologies. "
    "Yuyang frequently researches AI tools and techniques to enhance his development skills."
)

generated_queries = [
    "What are the best AI tools for developers?"
]

print("\nQuery Relevance Scores using BERTScore:")
for query in generated_queries:
    P, R, F1 = score([query], [persona_text], lang='en', verbose=True)

    print(f"Query: '{query}'")
    print(f"Precision: {P.mean().item():.4f}, Recall: {R.mean().item():.4f}, F1: {F1.mean().item():.4f}\n")



Query Relevance Scores using BERTScore:


tokenizer_config.json:   0%|          | 0.00/25.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/482 [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/899k [00:00<?, ?B/s]

merges.txt:   0%|          | 0.00/456k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.36M [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/1.42G [00:00<?, ?B/s]

Some weights of RobertaModel were not initialized from the model checkpoint at roberta-large and are newly initialized: ['roberta.pooler.dense.bias', 'roberta.pooler.dense.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


calculating scores...
computing bert embedding.


  0%|          | 0/1 [00:00<?, ?it/s]

computing greedy matching.


  0%|          | 0/1 [00:00<?, ?it/s]

done in 1.35 seconds, 0.74 sentences/sec
Query: 'What are the best AI tools for developers?'
Precision: 0.8782, Recall: 0.8329, F1: 0.8549



**Notes for resutls: **
- Precision: measures how much of the generated query is relevant to the persona text
- Recall: measures how much the persona text is covered by the generated query
- F1: harmonic mean of the precision and recall (F1 = 2 * ((precision * recall)/(precision + recall)))

In [ ]:
from bert_score import score

persona_references = [
    "Basketball",
    "Soccer",
    "Fashion",
    "Travel",
    "Technology"
]

generated_queries = [
    "What are the best AI tools for developers?",
    "What are some popular cat toys",
    "Give me some nice kitchen tools"
]

print("\nQuery relevance scores using BERTScore with multiple references:")
for query in generated_queries:
    P, R, F1 = score([query] * len(persona_references), persona_references, lang='en', verbose=True)

    # maximum F1 score across all references
    max_f1 = max(F1).item()
    print(f"Query: '{query}'\nMax F1 score across references: {max_f1:.4f}\n")



Query Relevance Scores using BERTScore with Multiple References:


Some weights of RobertaModel were not initialized from the model checkpoint at roberta-large and are newly initialized: ['roberta.pooler.dense.bias', 'roberta.pooler.dense.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


calculating scores...
computing bert embedding.


  0%|          | 0/1 [00:00<?, ?it/s]

computing greedy matching.


  0%|          | 0/1 [00:00<?, ?it/s]

done in 1.43 seconds, 3.50 sentences/sec
Query: 'What are the best AI tools for developers?'
Max F1 Score across references: 0.8155



Some weights of RobertaModel were not initialized from the model checkpoint at roberta-large and are newly initialized: ['roberta.pooler.dense.bias', 'roberta.pooler.dense.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


calculating scores...
computing bert embedding.


  0%|          | 0/1 [00:00<?, ?it/s]

computing greedy matching.


  0%|          | 0/1 [00:00<?, ?it/s]

done in 1.11 seconds, 4.50 sentences/sec
Query: 'What are some popular cat toys'
Max F1 Score across references: 0.8285



Some weights of RobertaModel were not initialized from the model checkpoint at roberta-large and are newly initialized: ['roberta.pooler.dense.bias', 'roberta.pooler.dense.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


calculating scores...
computing bert embedding.


  0%|          | 0/1 [00:00<?, ?it/s]

computing greedy matching.


  0%|          | 0/1 [00:00<?, ?it/s]

done in 0.88 seconds, 5.68 sentences/sec
Query: 'Give me some nice kitchen tools'
Max F1 Score across references: 0.8268



In [ ]:
#back to sentence bert again, since bertscore doesnt work well :(
from sentence_transformers import SentenceTransformer, util

model = SentenceTransformer('all-MiniLM-L6-v2')

persona_keywords = ["basketball", "soccer", "fashion", "travel", "technology"]

generated_queries = [
    "What are the best AI tools for developers?",
    "What is the newest released Nike shoe?",
    "How to cook pasta",
    "Top travel destinations for summer",
    "How to improve basketball shooting skills"
]

persona_embeddings = [model.encode(keyword, convert_to_tensor=True) for keyword in persona_keywords]

print("\nQuery Relevance Scores using BERT for Topic Matching:")
for query in generated_queries:
    query_embedding = model.encode(query, convert_to_tensor=True)

    max_similarity = 0
    matched_keyword = None
    for keyword, persona_embedding in zip(persona_keywords, persona_embeddings):
        similarity_score = util.cos_sim(query_embedding, persona_embedding).item()
        if similarity_score > max_similarity:
            max_similarity = similarity_score
            matched_keyword = keyword

    threshold = 0.5
    relevance = "Relevant" if max_similarity > threshold else "Not Relevant"
    print(f"Query: '{query}'\nMatched Keyword: {matched_keyword}\nMax Similarity Score: {max_similarity:.4f}\nRelevance: {relevance}\n")



Query Relevance Scores using BERT for Topic Matching:
Query: 'What are the best AI tools for developers?'
Matched Keyword: basketball
Max Similarity Score: 0.1307
Relevance: Not Relevant

Query: 'What is the newest released Nike shoe?'
Matched Keyword: fashion
Max Similarity Score: 0.3175
Relevance: Not Relevant

Query: 'How to cook pasta'
Matched Keyword: travel
Max Similarity Score: 0.0498
Relevance: Not Relevant

Query: 'Top travel destinations for summer'
Matched Keyword: travel
Max Similarity Score: 0.4709
Relevance: Not Relevant

Query: 'How to improve basketball shooting skills'
Matched Keyword: basketball
Max Similarity Score: 0.4331
Relevance: Not Relevant



In [ ]:
from transformers import pipeline

# zero-shot classification model
classifier = pipeline("zero-shot-classification", model="facebook/bart-large-mnli")

persona_labels = [
    "AI tools, software development, and technology",
    "clothing, shoes, and fashion trends",
    "cooking and recipes",
    "travel destinations and tourism",
    "basketball and sports training"
]


generated_queries = [
    "What are the best AI tools for developers?",
    "What is the newest released Nike shoe?",
    "How to cook pasta",
    "Top travel destinations for summer",
    "How to improve basketball shooting skills"
]

print("\nQuery Relevance using Zero-Shot Classification with Descriptive Labels:")
for query in generated_queries:
    result = classifier(query, candidate_labels=persona_labels)

    best_match = result['labels'][0]
    best_score = result['scores'][0]

    threshold = 0.6
    relevance = "Relevant" if best_score > threshold else "Not Relevant"

    print(f"Query: '{query}'")
    print(f"Extracted Topic: {best_match}, Score: {best_score:.4f}")
    print(f"Relevance: {relevance}\n")



Query Relevance using Zero-Shot Classification with Descriptive Labels:
Query: 'What are the best AI tools for developers?'
Extracted Topic: AI tools, software development, and technology, Score: 0.8897
Relevance: Relevant

Query: 'What is the newest released Nike shoe?'
Extracted Topic: clothing, shoes, and fashion trends, Score: 0.7092
Relevance: Relevant

Query: 'How to cook pasta'
Extracted Topic: cooking and recipes, Score: 0.9914
Relevance: Relevant

Query: 'Top travel destinations for summer'
Extracted Topic: travel destinations and tourism, Score: 0.9688
Relevance: Relevant

Query: 'How to improve basketball shooting skills'
Extracted Topic: basketball and sports training, Score: 0.9641
Relevance: Relevant



Another approach...
- Extract specific key phrases from the query.
- Define a set of broad topics (like "fashion", "technology", etc.).
- Use Sentence-BERT to convert both the key phrases and the broad topics into embeddings.
- Calculate the cosine similarity between the extracted key phrases and the broad topics to classify them.

In [ ]:
!pip install keybert


In [ ]:
from keybert import KeyBERT
from sentence_transformers import SentenceTransformer, util

kw_model = KeyBERT('all-MiniLM-L6-v2')
sentence_model = SentenceTransformer('all-MiniLM-L6-v2')

persona_keywords = ["basketball", "soccer", "fashion", "travel", "technology"]

broad_topics = ["fashion", "sports", "technology", "cooking", "travel"]

query = "What is the newest released Nike shoe?"

key_phrases = kw_model.extract_keywords(query, keyphrase_ngram_range=(1, 2), stop_words='english', top_n=3)
extracted_phrases = [phrase[0] for phrase in key_phrases]
print(f"Extracted Key Phrases: {extracted_phrases}")

topic_embeddings = {topic: sentence_model.encode(topic, convert_to_tensor=True) for topic in broad_topics}
phrase_embeddings = {phrase: sentence_model.encode(phrase, convert_to_tensor=True) for phrase in extracted_phrases}

classified_topics = []
for phrase, phrase_embedding in phrase_embeddings.items():
    max_similarity = 0
    best_topic = None
    for topic, topic_embedding in topic_embeddings.items():
        similarity = util.cos_sim(phrase_embedding, topic_embedding).item()
        if similarity > max_similarity:
            max_similarity = similarity
            best_topic = topic
    classified_topics.append(best_topic)

print(f"Generalized Topics for query '{query}': {classified_topics}")

is_relevant = any(topic in persona_keywords for topic in classified_topics)
print(f"Relevance: {'Relevant' if is_relevant else 'Not Relevant'}")


Extracted Key Phrases: ['nike shoe', 'released nike', 'nike']
Generalized Topics for query 'What is the newest released Nike shoe?': ['fashion', 'fashion', 'fashion']
Relevance: Relevant


Revised Approach

- Extract key phrases from the query using KeyBERT.
- Use semantic similarity between the extracted key phrases and the persona keywords to assign a relevance score.
- Calculate a similarity score between each extracted key phrase and the persona keywords directly, without needing any broad topic mapping.

In [ ]:
from keybert import KeyBERT
from sentence_transformers import SentenceTransformer, util
import numpy as np

kw_model = KeyBERT('all-MiniLM-L6-v2')
sentence_model = SentenceTransformer('all-MiniLM-L6-v2')

persona_keywords = ["basketball", "soccer", "fashion", "travel", "technology"]

query = "What is the newest released Nike shoe?"

key_phrases = kw_model.extract_keywords(query, keyphrase_ngram_range=(1, 2), stop_words='english', top_n=5)
extracted_phrases = [phrase[0] for phrase in key_phrases]
print(f"Extracted Key Phrases: {extracted_phrases}")

persona_embeddings = {keyword: sentence_model.encode(keyword, convert_to_tensor=True) for keyword in persona_keywords}
phrase_embeddings = {phrase: sentence_model.encode(phrase, convert_to_tensor=True) for phrase in extracted_phrases}

generalized_topics = []
for phrase, phrase_embedding in phrase_embeddings.items():
    max_similarity = 0
    best_match = None
    for keyword, persona_embedding in persona_embeddings.items():
        similarity = util.cos_sim(phrase_embedding, persona_embedding).item()
        if similarity > max_similarity:
            max_similarity = similarity
            best_match = keyword
    generalized_topics.append(best_match)

print(f"Generalized Topics for query '{query}': {generalized_topics}")

scores = []
for topic in generalized_topics:
    if topic:
        topic_embedding = sentence_model.encode(topic, convert_to_tensor=True)
        query_embedding = sentence_model.encode(query, convert_to_tensor=True)
        similarity_score = util.cos_sim(query_embedding, topic_embedding).item()
        scores.append(similarity_score)

if scores:
    average_score = np.mean(scores)
else:
    average_score = 0.0

print(f"Relevance Score for query '{query}': {average_score:.4f}")


Extracted Key Phrases: ['nike shoe', 'released nike', 'nike']
Relevance Score for query 'What is the newest released Nike shoe?': 0.4697


In [ ]:
from keybert import KeyBERT
from sentence_transformers import SentenceTransformer, util
from sklearn.cluster import AgglomerativeClustering
import numpy as np


kw_model = KeyBERT('all-MiniLM-L6-v2')
sentence_model = SentenceTransformer('all-MiniLM-L6-v2')

persona_keywords = ["AI", "basketball", "soccer", "fashion", "travel", "shoes", "technology"]


query = "What is the newest released Nike shoe?"

# extract key phrases from the query - KeyBERT
key_phrases = kw_model.extract_keywords(query, keyphrase_ngram_range=(1, 2), stop_words='english', top_n=5)
extracted_phrases = [phrase[0] for phrase in key_phrases]
print(f"Extracted Key Phrases: {extracted_phrases}")

# convert key phrases and persona keywords to embeddings
all_phrases = extracted_phrases + persona_keywords
phrase_embeddings = np.array([sentence_model.encode(phrase) for phrase in all_phrases])

# agglomerative clustering on the embeddings
clustering = AgglomerativeClustering(n_clusters=None, distance_threshold=1.0, metric='cosine', linkage='average')
cluster_labels = clustering.fit_predict(phrase_embeddings)

# group phrases by clusters
clusters = {}
for idx, label in enumerate(cluster_labels):
    if label not in clusters:
        clusters[label] = []
    clusters[label].append(all_phrases[idx])

# check if any cluster contains both a query phrase and a persona keyword
relevant = False
for cluster in clusters.values():
    has_query_phrase = any(phrase in extracted_phrases for phrase in cluster)
    has_persona_keyword = any(phrase in persona_keywords for phrase in cluster)
    if has_query_phrase and has_persona_keyword:
        relevant = True
        print(f"Matching Cluster: {cluster}")
        break

print(f"Relevance: {'Relevant' if relevant else 'Not Relevant'}")


Extracted Key Phrases: ['nike shoe', 'released nike', 'nike', 'newest released', 'shoe']
Matching Cluster: ['nike shoe', 'released nike', 'nike', 'newest released', 'shoe', 'AI', 'basketball', 'soccer', 'fashion', 'travel', 'shoes', 'technology']
Relevance: Relevant


In [ ]:
!pip install keybert nltk sentence-transformers


In [ ]:
import nltk
from nltk.corpus import wordnet as wn
from keybert import KeyBERT
from sentence_transformers import SentenceTransformer, util

nltk.download('wordnet')
kw_model = KeyBERT('all-MiniLM-L6-v2')
sentence_model = SentenceTransformer('all-MiniLM-L6-v2')

persona_keywords = ["AI", "basketball", "soccer", "fashion", "travel", "shoes", "technology"]


query = "What is the newest released Nike shoe?"

# extract key phrases from the query using KeyBERT
key_phrases = kw_model.extract_keywords(query, keyphrase_ngram_range=(1, 2), stop_words='english', top_n=5)
extracted_phrases = [phrase[0] for phrase in key_phrases]
print(f"Extracted Key Phrases: {extracted_phrases}")

# broaden terms - WordNet
def broaden_term_with_wordnet(term):
    hypernyms = set()
    synsets = wn.synsets(term)
    for synset in synsets:
        for hypernym in synset.hypernyms():
            hypernyms.add(hypernym.lemma_names()[0])
    return list(hypernyms)

broadened_query_topics = set()
for phrase in extracted_phrases:
    words = phrase.split()
    for word in words:
        broader_terms = broaden_term_with_wordnet(word)
        broadened_query_topics.update(broader_terms)

print(f"Broadened Query Topics: {broadened_query_topics}")

broadened_persona_topics = set()
for keyword in persona_keywords:
    broader_terms = broaden_term_with_wordnet(keyword)
    broadened_persona_topics.update(broader_terms)
broadened_persona_topics.update(persona_keywords)

print(f"Broadened Persona Topics: {broadened_persona_topics}")

# convert the broadened lists to embeddings and compute cosine similarity
query_embeddings = [sentence_model.encode(topic, convert_to_tensor=True) for topic in broadened_query_topics]
persona_embeddings = [sentence_model.encode(topic, convert_to_tensor=True) for topic in broadened_persona_topics]

# check for similarity
relevant = False
for query_emb in query_embeddings:
    for persona_emb in persona_embeddings:
        similarity = util.cos_sim(query_emb, persona_emb).item()
        if similarity > 0.6:
            relevant = True
            break
    if relevant:
        break
print(f"Relevance: {'Relevant' if relevant else 'Not Relevant'}")


/usr/local/lib/python3.10/dist-packages/sentence_transformers/cross_encoder/CrossEncoder.py:13: TqdmExperimentalWarning: Using `tqdm.autonotebook.tqdm` in notebook mode. Use `tqdm.tqdm` instead to force console mode (e.g. in jupyter console)
  from tqdm.autonotebook import tqdm, trange
[nltk_data] Downloading package wordnet to /root/nltk_data...
/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


modules.json:   0%|          | 0.00/349 [00:00<?, ?B/s]

config_sentence_transformers.json:   0%|          | 0.00/116 [00:00<?, ?B/s]

README.md:   0%|          | 0.00/10.7k [00:00<?, ?B/s]

sentence_bert_config.json:   0%|          | 0.00/53.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/612 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/90.9M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/350 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/466k [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/112 [00:00<?, ?B/s]

1_Pooling/config.json:   0%|          | 0.00/190 [00:00<?, ?B/s]

Extracted Key Phrases: ['nike shoe', 'released nike', 'nike', 'newest released', 'shoe']
Broadened Query Topics: {'issue', 'plate', 'case', 'footwear', 'restraint', 'transmit', 'exude', 'pass', 'generate', 'publicize', 'dress'}
Broadened Persona Topics: {'court_game', 'plate', 'case', 'technology', 'basketball', 'motion', 'practice', 'property', 'shoes', 'AI', 'basketball_equipment', 'dress', 'computer_science', 'soccer', 'vogue', 'restraint', 'ball', 'agency', 'insemination', 'discipline', 'situation', 'application', 'travel', 'movement', 'profession', 'footwear', 'consumer_goods', 'make', 'football', 'sloth', 'fashion'}
Relevance: Relevant


In [ ]:
import nltk
from nltk.corpus import wordnet as wn
from keybert import KeyBERT
from sentence_transformers import SentenceTransformer, util

nltk.download('wordnet')
kw_model = KeyBERT('all-MiniLM-L6-v2')
sentence_model = SentenceTransformer('all-MiniLM-L6-v2')

# persona keywords (input that won't be modified)
persona_keywords = ["AI", "basketball", "soccer", "fashion", "travel", "shoes", "technology"]


query = "What is the newest released Nike shoe?"

# extract key phrases from the query using KeyBERT
key_phrases = kw_model.extract_keywords(query, keyphrase_ngram_range=(1, 2), stop_words='english', top_n=5)
extracted_phrases = [phrase[0] for phrase in key_phrases]
print(f"Extracted Key Phrases: {extracted_phrases}\n")

# broaden terms using WordNet
def broaden_term_with_wordnet(term):
    hypernyms = set()
    synsets = wn.synsets(term)
    for synset in synsets:
        for hypernym in synset.hypernyms():
            hypernyms.add(hypernym.lemma_names()[0])
    return list(hypernyms)

# broaden the extracted key phrases
broadened_query_topics = set()
for phrase in extracted_phrases:
    words = phrase.split()
    for word in words:
        broader_terms = broaden_term_with_wordnet(word)
        print(f"Broadened terms for '{word}': {broader_terms}")
        broadened_query_topics.update(broader_terms)

print(f"\nBroadened Query Topics: {broadened_query_topics}\n")

# broaden persona keywords
broadened_persona_topics = set()
for keyword in persona_keywords:
    broader_terms = broaden_term_with_wordnet(keyword)
    print(f"Broadened terms for persona keyword '{keyword}': {broader_terms}")
    broadened_persona_topics.update(broader_terms)
broadened_persona_topics.update(persona_keywords)

print(f"\nBroadened Persona Topics: {broadened_persona_topics}\n")

# convert the broadened lists to embeddings
query_embeddings = {topic: sentence_model.encode(topic, convert_to_tensor=True) for topic in broadened_query_topics}
persona_embeddings = {topic: sentence_model.encode(topic, convert_to_tensor=True) for topic in broadened_persona_topics}

# check for cosine sim
print("Similarity Scores:")
relevant = False
for query_topic, query_emb in query_embeddings.items():
    for persona_topic, persona_emb in persona_embeddings.items():
        similarity = util.cos_sim(query_emb, persona_emb).item()
        print(f"Similarity between '{query_topic}' and '{persona_topic}': {similarity:.4f}")

        if similarity > 0.6:
            print(f"Matched: '{query_topic}' with Persona Topic: '{persona_topic}' (Similarity: {similarity:.4f})")
            relevant = True

    if relevant:
        break

print(f"\nRelevance: {'Relevant' if relevant else 'Not Relevant'}")


[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!


Extracted Key Phrases: ['nike shoe', 'released nike', 'nike', 'newest released', 'shoe']

Broadened terms for 'nike': []
Broadened terms for 'shoe': ['plate', 'case', 'footwear', 'restraint', 'dress']
Broadened terms for 'released': ['issue', 'transmit', 'exude', 'pass', 'generate', 'publicize']
Broadened terms for 'nike': []
Broadened terms for 'nike': []
Broadened terms for 'newest': []
Broadened terms for 'released': ['issue', 'transmit', 'exude', 'pass', 'generate', 'publicize']
Broadened terms for 'shoe': ['plate', 'case', 'footwear', 'restraint', 'dress']

Broadened Query Topics: {'issue', 'plate', 'case', 'footwear', 'restraint', 'transmit', 'exude', 'pass', 'generate', 'publicize', 'dress'}

Broadened terms for persona keyword 'AI': ['sloth', 'insemination', 'agency', 'computer_science']
Broadened terms for persona keyword 'basketball': ['court_game', 'basketball_equipment', 'ball']
Broadened terms for persona keyword 'soccer': ['football']
Broadened terms for persona keyword '

In [ ]:
!pip install scikit-learn --upgrade



In [ ]:
from keybert import KeyBERT
from sentence_transformers import SentenceTransformer, util
import numpy as np

kw_model = KeyBERT('all-MiniLM-L6-v2')
sentence_model = SentenceTransformer('all-MiniLM-L6-v2')

persona_keywords = ["AI", "sports", "fashion", "travel"]

query = "What is the newest released Nike shoe?"

# extract key phrases using KeyBERT
key_phrases = kw_model.extract_keywords(query, keyphrase_ngram_range=(1, 2), stop_words='english', top_n=5)
extracted_phrases = [phrase[0] for phrase in key_phrases]
print(f"Extracted Key Phrases: {extracted_phrases}\n")

# convert key phrases and persona keywords into embeddings
extracted_embeddings = {phrase: sentence_model.encode(phrase, convert_to_tensor=True) for phrase in extracted_phrases}
persona_embeddings = {keyword: sentence_model.encode(keyword, convert_to_tensor=True) for keyword in persona_keywords}

# not using WordNet..
# find related terms using cosine similarity within the extracted phrases and persona interests
def find_similar_terms(embedding, all_embeddings, top_n=3):
    similarities = []
    for term, emb in all_embeddings.items():
        similarity = util.cos_sim(embedding, emb).item()
        similarities.append((term, similarity))
    # Sort by similarity and return the top n terms
    similarities.sort(key=lambda x: x[1], reverse=True)
    return similarities[:top_n]

# broaden key phrases using embeddings
broadened_query_topics = set()
for phrase, phrase_embedding in extracted_embeddings.items():
    similar_terms = find_similar_terms(phrase_embedding, persona_embeddings, top_n=3)
    for term, score in similar_terms:
        broadened_query_topics.add(term)
    print(f"Broadened terms for '{phrase}': {[term for term, score in similar_terms]}")

print(f"\nBroadened Query Topics: {broadened_query_topics}\n")

# broaden the persona keywords using embeddings
broadened_persona_topics = set()
for keyword, persona_embedding in persona_embeddings.items():
    similar_terms = find_similar_terms(persona_embedding, extracted_embeddings, top_n=3)
    for term, score in similar_terms:
        broadened_persona_topics.add(term)
    print(f"Broadened terms for persona keyword '{keyword}': {[term for term, score in similar_terms]}")
broadened_persona_topics.update(persona_keywords)

print(f"\nBroadened Persona Topics: {broadened_persona_topics}\n")

print("Similarity Scores:")
relevant = False
for query_topic in broadened_query_topics:
    for persona_topic in broadened_persona_topics:
        query_emb = sentence_model.encode(query_topic, convert_to_tensor=True)
        persona_emb = sentence_model.encode(persona_topic, convert_to_tensor=True)
        similarity = util.cos_sim(query_emb, persona_emb).item()
        print(f"Similarity between '{query_topic}' and '{persona_topic}': {similarity:.4f}")
        if similarity > 0.6:
            print(f"Matched: '{query_topic}' with Persona Topic: '{persona_topic}' (Similarity: {similarity:.4f})")
            relevant = True
            break
    if relevant:
        break

print(f"\nRelevance: {'Relevant' if relevant else 'Not Relevant'}")


Extracted Key Phrases: ['nike shoe', 'released nike', 'nike', 'newest released', 'shoe']

Broadened terms for 'nike shoe': ['fashion', 'sports', 'AI']
Broadened terms for 'released nike': ['fashion', 'sports', 'AI']
Broadened terms for 'nike': ['fashion', 'sports', 'AI']
Broadened terms for 'newest released': ['fashion', 'sports', 'AI']
Broadened terms for 'shoe': ['fashion', 'sports', 'AI']

Broadened Query Topics: {'AI', 'fashion', 'sports'}

Broadened terms for persona keyword 'AI': ['shoe', 'nike', 'nike shoe']
Broadened terms for persona keyword 'sports': ['nike', 'shoe', 'nike shoe']
Broadened terms for persona keyword 'fashion': ['shoe', 'nike', 'nike shoe']
Broadened terms for persona keyword 'travel': ['shoe', 'nike', 'released nike']

Broadened Persona Topics: {'nike shoe', 'sports', 'nike', 'shoe', 'travel', 'fashion', 'released nike', 'AI'}

Similarity Scores:
Similarity between 'AI' and 'nike shoe': 0.2717
Similarity between 'AI' and 'sports': 0.3440
Similarity between 'AI

In [ ]:
!pip install sentence-transformers

In [ ]:
from keybert import KeyBERT
from sentence_transformers import SentenceTransformer, util
from sklearn.neighbors import NearestNeighbors
import numpy as np

kw_model = KeyBERT('all-MiniLM-L6-v2')
sentence_model = SentenceTransformer('all-mpnet-base-v2')

persona_keywords = ["AI", "sports", "fashion", "travel"]

query = "What is the newest released Nike shoe?"

# extract key phrases using KeyBERT
key_phrases = kw_model.extract_keywords(query, keyphrase_ngram_range=(1, 2), stop_words='english', top_n=5)
extracted_phrases = [phrase[0] for phrase in key_phrases]
print(f"Extracted Key Phrases: {extracted_phrases}\n")

# embeddings for extracted key phrases and persona keywords
extracted_embeddings = [sentence_model.encode(phrase) for phrase in extracted_phrases]
persona_embeddings = [sentence_model.encode(keyword) for keyword in persona_keywords]

def find_nearest_terms(embeddings, all_phrases, n_neighbors=3):
    nn_model = NearestNeighbors(n_neighbors=n_neighbors, metric='cosine')
    nn_model.fit(all_phrases)
    distances, indices = nn_model.kneighbors(embeddings)
    return indices

all_terms = extracted_phrases + persona_keywords
all_embeddings = [sentence_model.encode(term) for term in all_terms]

print("Broadened Query Topics:")
for i, embedding in enumerate(extracted_embeddings):
    indices = find_nearest_terms([embedding], all_embeddings)
    related_terms = [all_terms[idx] for idx in indices[0]]
    print(f"Broadened terms for '{extracted_phrases[i]}': {related_terms}")

print("\nBroadened Persona Topics:")
for i, embedding in enumerate(persona_embeddings):
    indices = find_nearest_terms([embedding], all_embeddings)
    related_terms = [all_terms[idx] for idx in indices[0]]
    print(f"Broadened terms for persona keyword '{persona_keywords[i]}': {related_terms}")

broadened_query_topics = set(extracted_phrases)
broadened_persona_topics = set(persona_keywords)

relevant = False
for query_term in broadened_query_topics:
    for persona_term in broadened_persona_topics:
        query_emb = sentence_model.encode(query_term)
        persona_emb = sentence_model.encode(persona_term)
        similarity = util.cos_sim(query_emb, persona_emb).item()
        print(f"Similarity between '{query_term}' and '{persona_term}': {similarity:.4f}")
        if similarity > 0.6:
            print(f"Matched: '{query_term}' with Persona Topic: '{persona_term}' (Similarity: {similarity:.4f})")
            relevant = True
            break
    if relevant:
        break

print(f"\nRelevance: {'Relevant' if relevant else 'Not Relevant'}")


modules.json:   0%|          | 0.00/349 [00:00<?, ?B/s]

config_sentence_transformers.json:   0%|          | 0.00/116 [00:00<?, ?B/s]

README.md:   0%|          | 0.00/10.6k [00:00<?, ?B/s]

sentence_bert_config.json:   0%|          | 0.00/53.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/571 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/438M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/363 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/466k [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/239 [00:00<?, ?B/s]

1_Pooling/config.json:   0%|          | 0.00/190 [00:00<?, ?B/s]

Extracted Key Phrases: ['nike shoe', 'released nike', 'nike', 'newest released', 'shoe']

Broadened Query Topics:
Broadened terms for 'nike shoe': ['nike shoe', 'nike', 'shoe']
Broadened terms for 'released nike': ['released nike', 'nike', 'nike shoe']
Broadened terms for 'nike': ['nike', 'nike shoe', 'released nike']
Broadened terms for 'newest released': ['newest released', 'released nike', 'nike shoe']
Broadened terms for 'shoe': ['shoe', 'nike shoe', 'nike']

Broadened Persona Topics:
Broadened terms for persona keyword 'AI': ['AI', 'fashion', 'sports']
Broadened terms for persona keyword 'sports': ['sports', 'nike', 'fashion']
Broadened terms for persona keyword 'fashion': ['fashion', 'nike', 'sports']
Broadened terms for persona keyword 'travel': ['travel', 'sports', 'fashion']
Similarity between 'nike shoe' and 'travel': 0.1971
Similarity between 'nike shoe' and 'fashion': 0.3357
Similarity between 'nike shoe' and 'AI': 0.1661
Similarity between 'nike shoe' and 'sports': 0.3639


In [ ]:
import requests
from keybert import KeyBERT
from sentence_transformers import SentenceTransformer, util

kw_model = KeyBERT('all-MiniLM-L6-v2')
sentence_model = SentenceTransformer('all-mpnet-base-v2')

persona_keywords = ["AI", "sports", "fashion", "travel"]

query = "What is the newest released Nike shoe?"

key_phrases = kw_model.extract_keywords(query, keyphrase_ngram_range=(1, 2), stop_words='english', top_n=5)
extracted_phrases = [phrase[0] for phrase in key_phrases]
print(f"Extracted Key Phrases: {extracted_phrases}\n")

def broaden_with_conceptnet(term):
    url = f"http://api.conceptnet.io/c/en/{term}?limit=5"
    response = requests.get(url)
    related_terms = set()
    if response.status_code == 200:
        data = response.json()
        for edge in data.get('edges', []):
            related_term = edge['end']['label'].lower()
            if related_term != term:
                related_terms.add(related_term)
    return list(related_terms)

broadened_query_topics = set()
for phrase in extracted_phrases:
    words = phrase.split()
    for word in words:
        broader_terms = broaden_with_conceptnet(word)
        print(f"Broadened terms for '{word}': {broader_terms}")
        broadened_query_topics.update(broader_terms)

print(f"\nBroadened Query Topics: {broadened_query_topics}\n")

query_embeddings = {topic: sentence_model.encode(topic, convert_to_tensor=True) for topic in broadened_query_topics}
persona_embeddings = {keyword: sentence_model.encode(keyword, convert_to_tensor=True) for keyword in persona_keywords}

print("Similarity Scores:")
relevant = False
for query_topic, query_emb in query_embeddings.items():
    for persona_topic, persona_emb in persona_embeddings.items():
        similarity = util.cos_sim(query_emb, persona_emb).item()
        print(f"Similarity between '{query_topic}' and '{persona_topic}': {similarity:.4f}")
        if similarity > 0.6:
            print(f"Matched: '{query_topic}' with Persona Topic: '{persona_topic}' (Similarity: {similarity:.4f})")
            relevant = True
            break
    if relevant:
        break

print(f"\nRelevance: {'Relevant' if relevant else 'Not Relevant'}")


Extracted Key Phrases: ['nike shoe', 'released nike', 'nike', 'newest released', 'shoe']

Broadened terms for 'nike': []
Broadened terms for 'shoe': ['footwear', 'foot', 'a closet', 'foot protection']
Broadened terms for 'released': []
Broadened terms for 'nike': []
Broadened terms for 'nike': []
Broadened terms for 'newest': ['new']
Broadened terms for 'released': []
Broadened terms for 'shoe': ['footwear', 'foot', 'a closet', 'foot protection']

Broadened Query Topics: {'footwear', 'foot protection', 'new', 'foot', 'a closet'}

Similarity Scores:
Similarity between 'footwear' and 'AI': 0.1731
Similarity between 'footwear' and 'sports': 0.3404
Similarity between 'footwear' and 'fashion': 0.5109
Similarity between 'footwear' and 'travel': 0.2320
Similarity between 'foot protection' and 'AI': 0.0788
Similarity between 'foot protection' and 'sports': 0.1946
Similarity between 'foot protection' and 'fashion': 0.2233
Similarity between 'foot protection' and 'travel': 0.1254
Similarity betw

# Enzo's work

Using UniEval

In [ ]:
!pip install transformers
from transformers import AutoTokenizer, AutoModelForSequenceClassification
import torch
from transformers import AutoTokenizer, AutoModelForSeq2SeqLM

#tokenizer = AutoTokenizer.from_pretrained("MingZhong/unieval-dialog")
#model = AutoModelForSeq2SeqLM.from_pretrained("MingZhong/unieval-dialog").to('cuda')
tokenizer = AutoTokenizer.from_pretrained("MingZhong/unieval-sum")
model = AutoModelForSequenceClassification.from_pretrained("MingZhong/unieval-sum").to('cuda')

/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


tokenizer_config.json:   0%|          | 0.00/1.89k [00:00<?, ?B/s]

spiece.model:   0%|          | 0.00/792k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/2.42M [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/1.79k [00:00<?, ?B/s]

config.json:   0%|          | 0.00/631 [00:00<?, ?B/s]

pytorch_model.bin:   0%|          | 0.00/3.13G [00:00<?, ?B/s]

Some weights of T5ForSequenceClassification were not initialized from the model checkpoint at MingZhong/unieval-sum and are newly initialized: ['classification_head.dense.bias', 'classification_head.dense.weight', 'classification_head.out_proj.bias', 'classification_head.out_proj.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


In [ ]:
example_prompt = [
    "Portable Wrench Holder", "what do skin tags look like on the face", "fuel gauges",
    "horseshoe purse", "DELL LAPTOP DOCKING STATION TRIPLE MONITOR", "engagement rings",
    "carpet liquidators", "operation gridlock", "moen shower faucet repair", "3t in sneakers",
    "concords 11", "lifeproof", "xbox one +kinect +adapter", "roaring 20s women's swimwear"
]
source_text = f"""
  Zhifei is a 24-year-old male from toronto working as a university student.
  They have interests in Workout, Cars, Money. Recently, he have been searching online for products like winter hiring, mac studio m2 ultra chip,
  rog snow keyboard coding, recommendation of rog keyboard, huawei hiring, used iphone Toronto, used RTX4090 Toronto, AMG CLE53.
  Based on his interests, occupation, and recent searches, generate a list of 10 realistic product search queries they might make.
  Each query should relate to their interests or recent searches and be similar in style to '{example_prompt}'.
  """

generated_text = "high-performance laptops for coding"

inputs = tokenizer(source_text, generated_text, return_tensors="pt", truncation=True, padding=True).to('cuda')
with torch.no_grad():
    outputs = model(**inputs)
    scores = torch.softmax(outputs.logits, dim=1)
if scores.shape[1] == 2:  # If only 2 metrics are provided
    coherence_score = scores[0][0].item()
    relevance_score = scores[0][1].item()
    fluency_score = relevance_score  # Placeholder for fluency
    consistency_score = coherence_score  # Placeholder for consistency
else:
    coherence_score = scores[0][0].item()
    consistency_score = scores[0][1].item()
    fluency_score = scores[0][2].item()
    relevance_score = scores[0][3].item()

print(f"Coherence Score: {coherence_score}")
print(f"Consistency Score: {consistency_score}")
print(f"Fluency Score: {fluency_score}")
print(f"Relevance Score: {relevance_score}")

average_score = (coherence_score + consistency_score + fluency_score + relevance_score) / 4
print(f"Average UniEval Score: {average_score}")

model.safetensors:   0%|          | 0.00/3.13G [00:00<?, ?B/s]

Passing a tuple of `past_key_values` is deprecated and will be removed in Transformers v4.48.0. You should pass an instance of `EncoderDecoderCache` instead, e.g. `past_key_values=EncoderDecoderCache.from_legacy_cache(past_key_values)`.


Coherence Score: 0.21745893359184265
Consistency Score: 0.21745893359184265
Fluency Score: 0.7825410962104797
Relevance Score: 0.7825410962104797
Average UniEval Score: 0.5000000149011612


Use BARTScore

In [ ]:
from transformers import BartTokenizer, BartForConditionalGeneration
import torch

tokenizer = BartTokenizer.from_pretrained("facebook/bart-large-cnn")
model = BartForConditionalGeneration.from_pretrained("facebook/bart-large-cnn").to("cuda" if torch.cuda.is_available() else "cpu")

In [ ]:
example_prompt = [
    "Portable Wrench Holder", "what do skin tags look like on the face", "fuel gauges",
    "horseshoe purse", "DELL LAPTOP DOCKING STATION TRIPLE MONITOR", "engagement rings",
    "carpet liquidators", "operation gridlock", "moen shower faucet repair", "3t in sneakers",
    "concords 11", "lifeproof", "xbox one +kinect +adapter", "roaring 20s women's swimwear"
]
source_text = f"""
  Zhifei is a 24-year-old male from toronto working as a university student.
  They have interests in Workout, Cars, Money. Recently, he have been searching online for products like winter hiring, mac studio m2 ultra chip,
  rog snow keyboard coding, recommendation of rog keyboard, huawei hiring, used iphone Toronto, used RTX4090 Toronto, AMG CLE53.
  Based on his interests, occupation, and recent searches, generate a list of 10 realistic product search queries they might make.
  Each query should relate to their interests or recent searches and be similar in style to '{example_prompt}'.
  """

generated_text = "Refurbished iPhone Deals Toronto"

input_ids = tokenizer(source_text, return_tensors="pt").input_ids.to(model.device)
target_ids = tokenizer(generated_text, return_tensors="pt").input_ids.to(model.device)


with torch.no_grad():
    outputs = model(input_ids=input_ids, labels=target_ids)
    log_likelihood = -outputs.loss.item()  # Loss is negative log likelihood

print(f"BARTScore: {log_likelihood}")

BARTScore: -6.580176830291748


Use BERTScore

In [ ]:
!pip install bert-score
!pip install sympy==1.8
!pip install --upgrade transformers bert_score

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 61.1/61.1 kB 5.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.1/6.1 MB 73.8 MB/s eta 0:00:00
  Attempting uninstall: sympy
    Found existing installation: sympy 1.13.1
    Uninstalling sympy-1.13.1:
      Successfully uninstalled sympy-1.13.1
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
torch 2.5.1+cu121 requires sympy==1.13.1; python_version >= "3.9", but you have sympy 1.8 which is incompatible.


   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.2/6.2 MB 70.8 MB/s eta 0:00:00
  Attempting uninstall: sympy
    Found existing installation: sympy 1.8
    Uninstalling sympy-1.8:
ERROR: Operation cancelled by user


In [ ]:
from bert_score import score

example_prompt = [
    "Portable Wrench Holder", "what do skin tags look like on the face", "fuel gauges",
    "horseshoe purse", "DELL LAPTOP DOCKING STATION TRIPLE MONITOR", "engagement rings",
    "carpet liquidators", "operation gridlock", "moen shower faucet repair", "3t in sneakers",
    "concords 11", "lifeproof", "xbox one +kinect +adapter", "roaring 20s women's swimwear"
]
source_text = f"""
  Zhifei is a 24-year-old male from toronto working as a university student.
  They have interests in Workout, Cars, Money. Recently, he have been searching online for products like winter hiring, mac studio m2 ultra chip,
  rog snow keyboard coding, recommendation of rog keyboard, huawei hiring, used iphone Toronto, used RTX4090 Toronto, AMG CLE53.
  Based on his interests, occupation, and recent searches, generate a list of 10 realistic product search queries they might make.
  Each query should relate to their interests or recent searches and be similar in style to '{example_prompt}'.
  """

generated_text = "high-performance laptops for coding"

print("\nQuery Relevance Scores using BERTScore:")

P, R, F1 = score([generated_text], [source_text], lang='en', model_type="bert-base-uncased", verbose=True)

print(f"Query: '{generated_text}'")
print(f"Precision: {P.mean().item():.4f}, Recall: {R.mean().item():.4f}, F1: {F1.mean().item():.4f}\n")


Query Relevance Scores using BERTScore:


tokenizer_config.json:   0%|          | 0.00/48.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/570 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

In [ ]:
!pip install bert-score

In [ ]:
from bert_score import score

persona_references = [
    "Workout", "Cars", "Money",
    #Search history
    "winter hiring", "mac studio m2 ultra chip",
    "rog snow keyboard coding", "recommendation of rog keyboard", "huawei hiring",
    "used iphone Toronto", "used RTX4090 Toronto", "AMG CLE53"
]

generated_queries = [
    "top recommendations for ROG keyboards"



]

print("\nQuery relevance scores using BERTScore with multiple references:")
for query in generated_queries:
    P, R, F1 = score([query] * len(persona_references), persona_references, lang='en', verbose=True)

    # maximum F1 score across all references
    max_f1 = max(F1).item()
    print(f"Query: '{query}'\nMax F1 score across references: {max_f1:.4f}\n")


Query relevance scores using BERTScore with multiple references:


Some weights of RobertaModel were not initialized from the model checkpoint at roberta-large and are newly initialized: ['roberta.pooler.dense.bias', 'roberta.pooler.dense.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


calculating scores...
computing bert embedding.


  0%|          | 0/1 [00:00<?, ?it/s]

computing greedy matching.


  0%|          | 0/1 [00:00<?, ?it/s]

done in 0.13 seconds, 83.77 sentences/sec
Query: 'top recommendations for ROG keyboards'
Max F1 score across references: 0.8593



Anthony: Test if we can get a substantially low BERTScore on unrelated persona & query.

In [2]:
!pip install bert-score

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 61.1/61.1 kB 2.4 MB/s eta 0:00:00


In [6]:
from bert_score import score

persona_references = [
    "Workout", "Cars", "Money",
    "winter hiring", "mac studio m2 ultra chip",
    "rog snow keyboard coding", "recommendation of rog keyboard", "huawei hiring",
    "used iphone Toronto", "used RTX4090 Toronto", "AMG CLE53"
]

generated_queries = [
    "All week long we will be highlighting some incredible University of Waterloo entrepreneurs",
    "Dynamic programming is a computer programming technique",
    "Best selling portable Windows laptops",
    "High-quality piano keyboards for beginners"
]

for query in generated_queries:
    P, R, F1 = score([query] * len(persona_references), persona_references, lang='en', verbose=True)
    max_f1 = max(F1).item()
    print(f"Query: '{query}'\nMax F1 score across references: {max_f1:.4f}\n")

Some weights of RobertaModel were not initialized from the model checkpoint at roberta-large and are newly initialized: ['roberta.pooler.dense.bias', 'roberta.pooler.dense.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


calculating scores...
computing bert embedding.


  0%|          | 0/1 [00:00<?, ?it/s]

computing greedy matching.


  0%|          | 0/1 [00:00<?, ?it/s]

done in 3.01 seconds, 3.65 sentences/sec
Query: 'All week long we will be highlighting some incredible University of Waterloo entrepreneurs'
Max F1 score across references: 0.8318



Some weights of RobertaModel were not initialized from the model checkpoint at roberta-large and are newly initialized: ['roberta.pooler.dense.bias', 'roberta.pooler.dense.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


calculating scores...
computing bert embedding.


  0%|          | 0/1 [00:00<?, ?it/s]

computing greedy matching.


  0%|          | 0/1 [00:00<?, ?it/s]

done in 1.92 seconds, 5.74 sentences/sec
Query: 'Dynamic programming is a computer programming technique'
Max F1 score across references: 0.8506



Some weights of RobertaModel were not initialized from the model checkpoint at roberta-large and are newly initialized: ['roberta.pooler.dense.bias', 'roberta.pooler.dense.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


calculating scores...
computing bert embedding.


  0%|          | 0/1 [00:00<?, ?it/s]

computing greedy matching.


  0%|          | 0/1 [00:00<?, ?it/s]

done in 1.13 seconds, 9.75 sentences/sec
Query: 'Best selling portable Windows laptops'
Max F1 score across references: 0.8747



Some weights of RobertaModel were not initialized from the model checkpoint at roberta-large and are newly initialized: ['roberta.pooler.dense.bias', 'roberta.pooler.dense.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


calculating scores...
computing bert embedding.


  0%|          | 0/1 [00:00<?, ?it/s]

computing greedy matching.


  0%|          | 0/1 [00:00<?, ?it/s]

done in 1.24 seconds, 8.87 sentences/sec
Query: 'High-quality piano keyboards for beginners'
Max F1 score across references: 0.8718



### Anthony's Comment:
BERTScore gives scores over 0.85 even for unrelated queries (e.g. a short description of dynamic programming), and it seems BERTScore gives a score over 0.8 most of the time.
We might want to re-consider our use of BERTScore for generated-query evaluation.